backends/azure/azure.go

```diff
@@ -0,0 +1,482 @@
+	// https://github.com/Azure/azure-sdk-for-go/blob/e5c902ce7aca5aa0f4c7bb7e46c18c8fc91ad458/sdk/storage/azblob/blockblob/models.go#L264
+	Concurrency int `yaml:"concurrency"`
+
+	// Not loaded from YAML
+	Logger logr.Logger `yaml:"-"`
+}
+
+type Backend struct {
+	opt        Options
+	client     *azblob.Client
+	log        logr.Logger
+	markerName string
+
+	mu         sync.Mutex
+	lastMarker string
+	lastList   simpleblob.BlobList
+	lastTime   time.Time
+}
+
+func (o Options) Check() error {
+	if o.UseEnvCreds {
+		return nil
+	}
+
+	hasSecretsCreds := o.AccountName != "" && o.AccountKey != ""
+
+	if !hasSecretsCreds {
+		return fmt.Errorf("azure storage.options: account_name and account_key are required")
+	}
+
+	if o.Container == "" {
+		return fmt.Errorf("azure storage.options: container is required")
+	}
+	return nil
+}
+
+// New creates a new backend instance.
+// The lifetime of the context passed in must span the lifetime of the whole
+// backend instance, not just the init time, so do not set any timeout on it!
+func New(ctx context.Context, opt Options) (*Backend, error) {
+	if opt.InitTimeout == 0 {
+		opt.InitTimeout = DefaultInitTimeout
+	}
+
+	if opt.UpdateMarkerForceListInterval == 0 {
+		opt.UpdateMarkerForceListInterval = DefaultUpdateMarkerForceListInterval
+	}
+
+	if opt.Concurrency == 0 {
+		opt.Concurrency = DefaultConcurrency
+	}
+
+	if err := opt.Check(); err != nil {
+		return nil, err
+	}
+
+	log := opt.Logger
+	if log.GetSink() == nil {
+		log = logr.Discard()
+	}
+	log = log.WithName("azure")
+
+	var endpoint string
+
+	accountName := opt.AccountName
+
+	if opt.EndpointURL != "" {
+		endpoint = opt.EndpointURL
+	} else {
+		endpoint = fmt.Sprintf("https://%s.blob.core.windows.net", accountName)
+	}
+
+	var client *azblob.Client
+
+	// If UseEnvCreds is set, we will attempt to use the environment variables and the Azure service principle based `azidentity.NewDefaultAzureCredential()` method
+	// https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azidentity/README.md#service-principal-with-secret
+	if opt.UseEnvCreds {
+		// Test if the environment variables are set
+		_, ok := os.LookupEnv("AZURE_CLIENT_ID")
+		if !ok {
+			return nil, errors.New("AZURE_CLIENT_ID could not be found")
+		}
+
+		_, ok = os.LookupEnv("AZURE_TENANT_ID")
+		if !ok {
+			return nil, errors.New("AZURE_TENANT_ID could not be found")
+		}
+
+		_, ok = os.LookupEnv("AZURE_CLIENT_SECRET")
+		if !ok {
+			return nil, errors.New("AZURE_CLIENT_SECRET could not be found")
+		}
+
+		cred, err := azidentity.NewDefaultAzureCredential(nil)
+		if err != nil {
+			return nil, err
+		}
+
+		client, err = azblob.NewClient(endpoint, cred, nil)
+
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		cred, err := azblob.NewSharedKeyCredential(accountName, opt.AccountKey)
+		if err != nil {
+			return nil, err
+		}
+
+		client, err = azblob.NewClientWithSharedKeyCredential(endpoint, cred, nil)
+
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
To my (Azure-)untrained eye, this seems convoluted. From what I can glean from the `azidentity` documentation linked in the comment, and a quick look at further documentation of authentication for Azure Blob Storage, I'd prefer this to be setup differently. I think the default way to authenticate should be to use `azidentity.NewDefaultAzureCredential`, without any checking of environment variables up front to allow it to use other kinds of credentials as well. A selectable alternative should be the `azblob`-specific Shared Key credential.

Reply:
To clarify: I mean having a `UseSharedKey` (or `UseSharedKeyCredential`)  option instead of an  `UseEnvCreds`, and not checking the env vars because DefaultAzureCredential can do much more than just this env variant.